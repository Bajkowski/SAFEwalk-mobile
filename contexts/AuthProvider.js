import React, { createContext, useReducer, useMemo } from "react";
import { AsyncStorage } from "react-native";

export const AuthContext = createContext();

export function AuthProvider({ children }) {
  const [state, dispatch] = useReducer(authReducer, {
    // initial states
    isLoading: true,
    isSignout: false,
    userToken: null,
    userType: "user",
    userEmail: null
  });

  // Handle Login
  async function login(userType, { email, password }) {
    try {
      // Put code below for this: make login GET request to server

      // For now, while server is not set, use email as value for user token
      await AsyncStorage.setItem("userToken", JSON.stringify(email));
      // Store the user type i.e. which type of user is logged in, user or SAFEwalker.
      await AsyncStorage.setItem("userType", userType);
      // Store the user email
      await AsyncStorage.setItem("userEmail", JSON.stringify(email));

      dispatch({
        type: "LOG_IN",
        token: JSON.stringify(email),
        userType: userType,
        userEmail: email
      });
    } catch (error) {
      throw new Error("Error in login(): " + error);
    }
  }

  // Handle signout
  async function signout() {
    try {
      await AsyncStorage.removeItem("userToken");
      await AsyncStorage.removeItem("userType");
      await AsyncStorage.removeItem("userEmail");

      dispatch({ type: "SIGN_OUT" });
    } catch (error) {
      throw new Error("Error in signout(): " + error);
    }
  }

  // Handle register
  async function register({
    email,
    password,
    firstName,
    lastName,
    phoneNumber
  }) {
    try {
      // Put code below for this: make register POST request to server

      // For now, while server is not set, use email as value for user token
      await AsyncStorage.setItem("userToken", JSON.stringify(email));
      // Store the user type i.e. which type of user is logged in, user or SAFEwalker.
      await AsyncStorage.setItem("userType", userType);
      // Store the user email
      await AsyncStorage.setItem("userEmail", JSON.stringify(email));

      dispatch({
        type: "LOG_IN",
        token: JSON.stringify(email),
        userType: "user",
        userEmail: email
      });
    } catch (error) {
      throw new Error("Error in register(): " + error);
    }
  }

  const authContextValue = useMemo(() => {
    return {
      isLoading: state.isLoading,
      isSignout: state.isSignout,
      userToken: state.userToken,
      userType: state.userType,
      userEmail: state.userEmail,
      dispatch,
      login,
      signout,
      register
    };
  }, [state]);

  return (
    <AuthContext.Provider value={authContextValue}>
      {children}
    </AuthContext.Provider>
  );
}

function authReducer(prevState, action) {
  switch (action.type) {
    case "RESTORE_TOKEN":
      return {
        ...prevState,
        isLoading: false,
        userToken: action.token,
        userType: action.userType,
        userEmail: action.userEmail
      };
    case "LOG_IN":
      return {
        ...prevState,
        isSignout: false,
        userToken: action.token,
        userType: action.userType,
        userEmail: action.userEmail
      };
    case "SIGN_OUT":
      return {
        ...prevState,
        isSignout: true,
        userToken: null,
        userType: null,
        userEmail: null
      };
    default: {
      throw new Error(`Unhandled auth action type: ${action.type}`);
    }
  }
}
