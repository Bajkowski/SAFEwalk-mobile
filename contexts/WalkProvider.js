import React, { createContext, useReducer, useMemo } from "react";
import { AsyncStorage } from "react-native";

export const WalkContext = createContext();

export function WalkProvider({ children }) {
  const [state, dispatch] = useReducer(walkReducer, {
    // initial states
    isWalkLoading: true,
    isWalkActive: "false", // has to be a string due to AsyncStorage constraints (cannot store boolean)

    // initial walk state
    walkId: null,

    // initial user state
    userEmail: null,
    walkerEmail: null,

    // initial SAFEwalker state
    userSocketId: null,
    walkerSocketId: null,
  });

  // Set up walk ID
  async function setWalkAsActive() {
    try {
      // Store walk ID
      await AsyncStorage.setItem("isWalkActive", "true");

      dispatch({
        type: "SET_WALK_ACTIVE",
        isWalkActive: "true",
      });
    } catch (error) {
      console.error("Error in setIsWalkActive(): " + error);
    }
  }

  // Set up walk ID
  async function setWalkId(walkId) {
    try {
      // Store walk ID
      await AsyncStorage.setItem("walkId", walkId);

      dispatch({
        type: "SET_WALK_ID",
        walkId: walkId,
      });
    } catch (error) {
      console.error("Error in setWalkId(): " + error);
    }
  }

  // Remove walk ID
  async function removeWalkId() {
    try {
      // remove walk ID
      await AsyncStorage.removeItem("WalkId");

      dispatch({ type: "REMOVE_WALK_ID" });
    } catch (error) {
      console.error("Error in removeWalkId(): " + error);
    }
  }

  // Store User information (walkId, email and socket ID)
  async function setUserInfo(walkId, userEmail, userSocketId) {
    try {
      await AsyncStorage.setItem("walkId", walkId);
      await AsyncStorage.setItem("userEmail", userEmail);
      await AsyncStorage.setItem("userSocketId", userSocketId);

      dispatch({
        type: "SET_USER_INFO",
        walkId: walkId,
        userEmail: userEmail,
        userSocketId: userSocketId,
      });
    } catch (error) {
      console.error("Error in seUserInfo(): " + error);
    }
  }

  // Store Safewalker information (email and socket ID)
  async function setWalkerInfo(walkerEmail, walkerSocketId) {
    try {
      await AsyncStorage.setItem("walkerEmail", walkerEmail);
      await AsyncStorage.setItem("walkerSocketId", walkerSocketId);

      dispatch({
        type: "SET_WALKER_INFO",
        walkerEmail: walkerEmail,
        walkerSocketId: walkerSocketId,
      });
    } catch (error) {
      console.error("Error in setWalkerInfo(): " + error);
    }
  }

  // remove all current walk-related information
  async function resetWalkContextState() {
    try {
      await AsyncStorage.removeItem("walkId");
      await AsyncStorage.removeItem("userEmail");
      await AsyncStorage.removeItem("walkerEmail");
      await AsyncStorage.removeItem("userSocketId");
      await AsyncStorage.removeItem("walkerSocketId");

      dispatch({ type: "RESET" });
    } catch (error) {
      console.error("Error in resetContextState(): " + error);
    }
  }

  const walkContextValue = useMemo(() => {
    return {
      // states
      isWalkLoading: state.isWalkLoading,
      isWalkActive: state.isWalkActive,
      walkId: state.walkId,
      userEmail: state.userEmail,
      walkerEmail: state.walkerEmail,
      userSocketId: state.userSocketId,
      walkerSocketId: state.walkerSocketId,

      // functions
      setWalkAsActive,
      setWalkId,
      removeWalkId,
      setUserInfo,
      setWalkerInfo,
      resetWalkContextState,
      dispatch,
    };
  }, [state]);

  return (
    <WalkContext.Provider value={walkContextValue}>
      {children}
    </WalkContext.Provider>
  );
}

function walkReducer(prevState, action) {
  switch (action.type) {
    case "RESTORE_STATE":
      return {
        ...prevState,
        isWalkLoading: false,
        isWalkActive: action.isWalkActive,
        walkId: action.walkId,
        userEmail: action.userEmail,
        walkerEmail: action.walkerEmail,
        userSocketId: action.userSocketId,
        walkerSocketId: action.walkerSocketId,
      };
    case "SET_WALK_ACTIVE":
      return {
        ...prevState,
        isWalkActive: "true",
      };
    case "SET_WALK_ID":
      return {
        ...prevState,
        walkId: action.walkId,
      };
    case "REMOVE_WALK_ID":
      return {
        ...prevState,
        walkId: null,
      };
    case "SET_USER_INFO":
      return {
        ...prevState,
        walkId: action.walkId,
        userEmail: action.userEmail,
        userSocketId: action.userSocketId,
      };
    case "SET_WALKER_INFO":
      return {
        ...prevState,
        walkerEmail: action.walkerEmail,
        walkerSocketId: action.walkerSocketId,
      };
    case "RESET":
      return {
        ...prevState,
        isWalkLoading: false,
        isWalkActive: "false",
        walkId: null,
        userEmail: null,
        userSocketId: null,
        walkerEmail: null,
        walkerSocketId: null,
      };

    default: {
      console.error(`Unhandled walk action type: ${action.type}`);
    }
  }
}
