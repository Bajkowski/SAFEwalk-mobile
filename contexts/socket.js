const io = require("socket.io-client");
module.exports = io("http://127.0.0.1:3000");
// module.exports = io("http://365b17d8.ngrok.io");
