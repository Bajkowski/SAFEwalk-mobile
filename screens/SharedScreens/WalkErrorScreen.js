import React, { useContext } from "react";
import { View, Button, StyleSheet } from "react-native";
import { Text } from "react-native-elements";

import { AuthContext } from "../../contexts/AuthProvider";

export default function WalkErrorScreen(props) {
  const { userType, signout } = useContext(AuthContext);
  function handlePress() {
    signout();
  }
  return (
    <View style={styles.container}>
      <View>
        {/* If userType is neither "user" nor "safewalker", there's something wrong */}
        <Text h1>Error</Text>
        <Text h3>userType is neither "user" nor "safewalker" in</Text>
        <Text h3>{props.filename}</Text>
        <Text h3>userType value:"{userType}"</Text>
        <Text h2>Please sign out.</Text>
        <Button title="Sign out" onPress={() => handlePress()} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignSelf: "center",
    justifyContent: "center",
    paddingHorizontal: 20,
  },
});
