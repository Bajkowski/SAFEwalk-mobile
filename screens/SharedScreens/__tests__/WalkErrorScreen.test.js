import React from "react";
import renderer from "react-test-renderer";
import { render, act, fireEvent } from "@testing-library/react-native";

import WalkErrorScreen from "../WalkErrorScreen";
import { AuthProvider } from "../../../contexts/AuthProvider";

export const createTestProps = (props) => ({
  filename: "RandomScreen.js",
});

let props;
let component;
let screen;

describe("WalkErrorScreen ", () => {
  beforeEach(() => {
    props = createTestProps();

    component = (
      <AuthProvider>
        <WalkErrorScreen filename={props.filename} />
      </AuthProvider>
    );

    screen = render(component);
  });

  it("renders correctly", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("signs out upon button press", () => {
    let signoutButton = screen.getByLabelText("Sign out");
    expect(signoutButton).toBeTruthy();

    act(() => {
      fireEvent.press(signoutButton);
    });
  });
});
