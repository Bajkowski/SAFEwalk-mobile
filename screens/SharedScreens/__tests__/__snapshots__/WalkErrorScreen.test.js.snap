// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalkErrorScreen  renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontSize": 50,
        }
      }
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      Error
    </Text>
    <Text
      style={
        Object {
          "fontSize": 35,
        }
      }
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      userType is neither "user" nor "safewalker" in
    </Text>
    <Text
      style={
        Object {
          "fontSize": 35,
        }
      }
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      RandomScreen.js
    </Text>
    <Text
      style={
        Object {
          "fontSize": 35,
        }
      }
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      userType value:"
      user
      "
    </Text>
    <Text
      style={
        Object {
          "fontSize": 42.5,
        }
      }
      theme={
        Object {
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      Please sign out.
    </Text>
    <Button
      accessibilityLabel="Sign out"
      onPress={[Function]}
      title="Sign out"
    />
  </View>
</View>
`;
